File: src/PieChart/PieChartTypes.ts
export default interface PieChartParam {
  canvas: HTMLCanvasElement;
  chartType?: 'primary' | 'donut';
  data: PieChartData[];
  chartFillColor?: string;
  chartSize?: number;
  totalValue?: number;
  displayValue?: {
    visible?: boolean;
    color?: string;
    style?: string;
  };
}

export interface TextType {
  text: any | any[];
  visible?: boolean;
  color?: string;
  style?: string;
}

interface HoverType {
  chartColor: string;
  fontColor?: string;
  fontStyle?: string;
}

interface PieChartData {
  title?: TextType;
  value: number;
  fillColor: string;
  hover?: HoverType;
}

interface PieChartElement extends PieChartData {
  angleValue: number;
}

interface EachDataAreaType {
  title?: TextType;
  startEndDegree: number;
  hover?: HoverType;
}

interface PieChartType {
  readonly PRIMARY: string;
  readonly DONUT: string;
}

const pieChartType: PieChartType = {
  PRIMARY: 'primary',
  DONUT: 'donut',
};

interface DefaultChartValue {
  readonly CHART_COLOR: string;
  readonly HOVER_CHART_COLOR: string;
  readonly FONT_STYLE: string;
  readonly FONT_COLOR: string;
  readonly CHART_SIZE: number;
}

const defaultChartValue: DefaultChartValue = {
  CHART_COLOR: 'rgb(230, 230, 230)',
  HOVER_CHART_COLOR: '#BAAAD7',
  FONT_STYLE: 'normal bold 15px serif',
  FONT_COLOR: 'black',
  CHART_SIZE: 500,
};

interface PieChartEvent {
  result1: boolean;
  result2: boolean;
  index: number;
  degree: number;
}

export {
  TextType as DataTitle,
  PieChartData,
  PieChartElement,
  EachDataAreaType,
  PieChartType,
  DefaultChartValue,
  PieChartEvent,
  pieChartType,
  defaultChartValue,
};

File: wasm_module/src/pie_chart/pie_chart_types.rs
use serde::{Serialize, Deserialize};

#[derive(Clone, Serialize, Deserialize)]
pub struct HoverType {
  chart_color: String,
  font_color: Option<String>,
  font_style: Option<String>,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct TextType {
  text: String,
  visible: Option<bool>,
  color: Option<String>,
  style: Option<String>,
}

pub struct SizeType {
  width: i32,
  height: i32,
}

pub struct PositionType {
  x: i32,
  y: i32,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct PieChartData {
  title: Option<TextType>,
  value: f64,
  fill_color: String,
  hover: Option<HoverType>,
}

// #[derive(Clone, Serialize, Deserialize)]
pub struct PieChartElement {
  title: Option<TextType>,
  value: f64,
  fill_color: String,
  hover: Option<HoverType>,
  angle_value: f64,
}

pub struct PieChartEachArea {
  title: Option<TextType>,
  start_end_degree: f64,
  hover: Option<HoverType>
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DisplayValueType {
  visible: bool,
  color: String,
  style: String,
}